src/charm.py

```diff
@@ -0,0 +1,156 @@
+from ops.pebble import Layer
+
+from types_ import Credentials
+
+if TYPE_CHECKING:
+    from ops.pebble import LayerDict  # pragma: no cover
+
+logger = logging.getLogger(__name__)
+
+JENKINS_WEB_URL = "http://localhost:8080"
+JENKINS_HOME = Path("/var/lib/jenkins")
+# Path to initial admin password file
+INITIAL_PASSWORD = JENKINS_HOME / Path("secrets/initialAdminPassword")
+# Path to last executed jenkins version file, required to override wizard installation
+LAST_EXEC = JENKINS_HOME / Path("jenkins.install.InstallUtil.lastExecVersion")
+# Path to jenkins version file, required to override wizard installation
+UPDATE_VERSION = JENKINS_HOME / Path("jenkins.install.UpgradeWizard.state")
+
+
+class JenkinsK8SOperatorCharm(CharmBase):
+    """Charm Jenkins."""
+
+    def __init__(self, *args: Any):
+        """Initialize the charm and register event handlers.
+
+        Args:
+            args: Arguments to initialize the char base.
+        """
+        super().__init__(*args)
+        self.framework.observe(self.on.jenkins_pebble_ready, self._on_jenkins_pebble_ready)
+
+    def _get_pebble_layer(self, env: dict[str, str] | None = None) -> Layer:
+        """Return a dictionary representing a Pebble layer.
+
+        Args:
+            env: Map of Jenkins environment variables.
+
+        Returns:
+            The pebble layer defining Jenkins service layer.
+        """
+        default_env = {"JENKINS_HOME": str(JENKINS_HOME)}
+        merged_env = default_env | env if env else default_env
+        layer: LayerDict = {
+            "summary": "jenkins layer",
+            "description": "pebble config layer for jenkins",
+            "services": {
+                "jenkins": {
+                    "override": "replace",
+                    "summary": "jenkins",
+                    "command": "java -Djava.awt.headless=true -jar /srv/jenkins/jenkins.war",
+                    "startup": "enabled",
+                    "environment": merged_env,
+                }
+            },
+        }
+        return Layer(layer)
+
+    def _is_jenkins_ready(self) -> bool:
+        """Check if Jenkins webserver is ready.
+
+        Returns:
+            True if Jenkins server is online. False otherwise.
+        """
+        return requests.get(f"{JENKINS_WEB_URL}/login", timeout=10).ok
+
+    def _wait_jenkins_ready(self, timeout: int = 140, check_interval: int = 10) -> None:
+        """Wait until Jenkins service is up.
+
+        Args:
+            timeout: Time in seconds to wait for jenkins to become ready in 10 second intervals.
+            check_interval: Time in seconds to wait between ready checks.
+
+        Raises:
+            TimeoutError: if Jenkins status check did not pass within the timeout duration.
+        """
+        for _ in range(timeout // check_interval):
+            if self._is_jenkins_ready():
+                break
+            sleep(check_interval)
+        else:
```

Comment:
This is a minor feedback, the for loop ends up calling a final `sleep` if Jenkins never gets ready and then raises which seems to be a bit unexpected. As a operator I would expect a sleep to be followed by a check whether Jenkins is ready or not have the sleep. What do you think?